### ISSUE TITLE
isolatedDeclarations should not have warning for functions that have no return statements

### ISSUE THREAD

**Issue by Jack-Works** [microsoft/TypeScript#58330]
### 🔎 Search Terms

isolatedDeclaration

### 🕗 Version & Regression Information

5.5.0

### ⏯ Playground Link

https://www.typescriptlang.org/play/?isolatedDeclarations=true&ts=5.5.0-dev.20240426#code/KYDwDg9gTgLgBAMwK4DsDGMCWEWIBQCUcA3gL5A

### 💻 Code

```ts
export function f() {}
```

### 🙁 Actual behavior

Function must have an explicit return type annotation with --isolatedDeclarations. (9007)

### 🙂 Expected behavior

No error

### Additional information about the issue

Functions with no return statements can be easily inferred as type `void`, so there is no need to give this warning.

**Comment by jakebailey** [comment]
Not doing this was intentional; a function without a return may not always be `void`.

```ts
declare function fail(): never;
export function f() {
	fail();
}
export const g = () => { fail(); }
```

Is:

```ts
export declare function f(): void;
export declare const g: () => never;
```


[Playground Link](https://www.typescriptlang.org/play/?isolatedDeclarations=true&ts=5.5.0-dev.20240426#code/CYUwxgNghgTiAEAzArgOzAFwJYHtVKiwgAoBKALnlRADcQYBuAWACgQAPABxxgyTUy58iMvADerAJCJCJUsxYBfVh2694YPAGc+Ac3gBeeKIMA+cQSJkG8RUA)

**Comment by RyanCavanaugh** [comment]
Correct error for function expressions, but not function declarations - declarations won't infer `never`

**Comment by jakebailey** [comment]
#58331 is indiscriminate in what it affects; if we want to allow this for function declarations only, that may be possible, but it sure seems like an extra inconsistency just to avoid 5 characters...

**Comment by Jack-Works** [comment]
hmm, I didn't think of `never`, so the function expression part is wrong. But I find it annoying to write `: void` for obviously clear function declarations.

**Comment by RyanCavanaugh** [comment]
Discussed and we don't want to make it untenable to, in the future, change the rules around function declaration return type inference. For example, it's maybe preferable that
```ts
function dofail() {
  Debug.fail("oops");
}
```
have an inferred return type of `never` (as it would if written as a function expression) rather than `void`.
